mkosi/vmspawn.py

```diff
@@ -0,0 +1,227 @@
+        die("RuntimeTrees= cannot be used when booting in BIOS firmware")
+
+    if config.qemu_kvm == ConfigFeature.enabled and not QemuDeviceNode.kvm.available():
+        die("KVM acceleration requested but cannot access /dev/kvm")
+
+    if config.qemu_vsock == ConfigFeature.enabled and QemuDeviceNode.vhost_vsock not in qemu_device_fds:
+        die("VSock requested but cannot access /dev/vhost-vsock")
+
+    if config.qemu_cdrom:
+        die("systemd-vmspawn does not support CD-ROM images")
+
+    if config.qemu_kernel:
+        kernel = config.qemu_kernel
+    elif "-kernel" in args.cmdline:
+        kernel = Path(args.cmdline[args.cmdline.index("-kernel") + 1])
+    else:
+        kernel = None
+
+    if kernel and not kernel.exists():
+        die(f"Kernel not found at {kernel}")
+
+    if config.qemu_firmware == QemuFirmware.auto:
+        if kernel:
+            firmware = QemuFirmware.uefi if KernelType.identify(kernel) != KernelType.unknown else QemuFirmware.linux
+        elif (
+            config.output_format in (OutputFormat.cpio, OutputFormat.directory) or
+            config.architecture.to_efi() is None
+        ):
+            firmware = QemuFirmware.linux
+        else:
+            firmware = QemuFirmware.uefi
+    else:
+        firmware = config.qemu_firmware
+
+    if firmware == QemuFirmware.linux:
+        kernel = config.output_dir_or_cwd() / config.output_split_kernel
+        if not kernel.exists():
+            die(
+                f"Kernel or UKI not found at {kernel}, please install a kernel in the image "
+                "or provide a -kernel argument to mkosi vmspawn"
+            )
+
+    notifications: dict[str, str] = {}
+    cmdline: list[PathString] = [
+        "systemd-vmspawn",
+        f"--qemu-smp={config.qemu_smp}",
+        f"--qemu-mem={config.qemu_mem}",
+        f"--qemu-kvm={config_feature_to_tristate(config.qemu_kvm)}",
+
+        # we need to handle this ourselves as we cannot pass the fd to vmspawn
+        # directly but have to pass it to qemu instead
+        # to prevent a conflict we disable vsock for vmspawn
+        "--qemu-vsock=no",
+    ]
+
+    if config.qemu_gui:
+        cmdline.append("--qemu-gui")
+
+    if config.secure_boot:
+        cmdline.append("--secure-boot=yes")
+
+    for name, val in config.credentials.items():
+        cmdline.append(f"--set-credential={name}:{val}")
+
+    # values which we need to pass directly to qemu
+    qemu_cmdline: list[PathString] = []
+
+    if QemuDeviceNode.vhost_vsock in qemu_device_fds:
+        qemu_cmdline += [
+            "-device",
+            f"vhost-vsock-pci,guest-cid={machine_cid(config)},vhostfd={qemu_device_fds[QemuDeviceNode.vhost_vsock]}"
+        ]
+
+    with contextlib.ExitStack() as stack:
+        if config.ephemeral:
+            fname = stack.enter_context(copy_ephemeral(config, config.output_dir_or_cwd() / config.output))
+        else:
+            fname = config.output_dir_or_cwd() / config.output
```

Comment:
This needs to use `config.output_with_compression()`

Reply:
Fixed in 594d4fb.